### PR TITLE
greetings: regularly print a string to the console every second.

### PR DESCRIPTION
**Pull request by Flameeyes**
Introduce a new sample app to repeatedly and regularly print a string
("Zephyr says 'Hi!'") every second on the console.

Recommend this sample instead of hello_world for boards that are not
meeting the minimum requirements for blinky.

### CODE REVIEW

**Review by kartben — CHANGES_REQUESTED** [review]
Good idea. Could you please also update the README ("Overview" and "Sample Output")? Thanks!

File: samples/hello_world/src/main.c [hunk 1]
```diff
@@ -6,8 +6,14 @@
...
 
+/* delay between greetings (in ms) */
+#define SLEEPTIME 1000
+
 int main(void)
 {
-	printk("Hello World! %s\n", CONFIG_BOARD);
+	while(1) {
+		printk("Hello World! %s\n", CONFIG_BOARD);
+		k_msleep(SLEEPTIME);
```
Comment (gmarull) [inline samples/hello_world/src/main.c:16]:
```suggestion
		k_msleep(1000);
```

File: samples/hello_world/src/main.c [hunk 2]
```diff
@@ -6,8 +6,14 @@
 
 #include <zephyr/kernel.h>
 
+/* delay between greetings (in ms) */
+#define SLEEPTIME 1000
+
```
Comment (gmarull) [inline samples/hello_world/src/main.c:11]:
```suggestion
```

File: samples/hello_world/src/main.c (hunk 1 above)
Comment (gmarull) [reply samples/hello_world/src/main.c:16]: just for simplicity, also, I'd change `printk` to `printf`, and include `<stdio.h>`

**Review by nashif — DISMISSED** [review]
just use `samples/synchronization`/ sample. Lets not change the behavior of a classic hello world sample.

**Comment by Flameeyes** [comment]
> just use `samples/synchronization`/ sample. Lets not change the behavior of a classic hello world sample.

I did that originally, but that does increase the noise surface significantly, since it's involving threading (attach debugger, where are you?)

During bring up of a new board (particularly if, like me, you have no idea what you're doing and is the first time you try), being able to exclude as much of the higher level stack is useful. Already bringing up the serial console adds more moving parts, which is why at least for the local testing I preferred reducing the surface area.

Would a trade-off of having this `#define` controllable be worth exploring?

**Comment by fabiobaltieri** [comment]
I also think it's a good idea to have it repeat, then you'd still get an output if you are fiddling with serial ports and end up missing the boot message for whatever reason.

**Comment by nashif** [comment]
> I also think it's a good idea to have it repeat, then you'd still get an output if you are fiddling with serial ports and end up missing the boot message for whatever reason.

again, we have the syncronization sample for this :)

**Comment by kartben** [comment]
> > I also think it's a good idea to have it repeat, then you'd still get an output if you are fiddling with serial ports and end up missing the boot message for whatever reason.
> 
> again, we have the syncronization sample for this :)

TBH it is a bit of a stretch to compare improving the Hello World sample by adding a simple sleep/while(1) to a full blown synchronization sample that demonstrates how to create threads, use semaphores, etc. Should we refrain from using k_sleep in every sample then, for this should be something reserved only for the synchronization sample? :)

I think the point here is really about improving the usability/UX of the Hello World (in addition to also making it a quick proof point that clock is setup properly, to @Flameeyes' point earlier), and nothing more.

**Review by kartben — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]
I'm with Ben on this, `synchronization` is good for showing thread usage but that's few steps ahead and this PR made me realize how may times I've hacked a printk into `blinky` while searching for the TX pin on whatever board I had just for this.

Though it is indeed sad to diverge from the classic version of it, we were already a bit off from the start, it should have been a `printf` and all lowercase.

x

**Comment by nashif** [comment]
https://www.geeksforgeeks.org/hello-world-in-30-different-languages/

hello world is a defact "standard", it has been the same code in Zephyr since we started, lets not change that just because of someone's debug needs, if you want to debug something, take hello world and modify it.

**Comment by Flameeyes** [comment]
Let me try to find a way forward:

 * there seem to be a want, besides from me, for a simple, regular output on the UART for various debugging purposes;
 * what made sense to me, is the equivalent of blinky, that is to output something every second or so — I have done that on hello_world because that's what [the documentation](https://docs.zephyrproject.org/latest/develop/getting_started/index.html#build-the-blinky-sample) suggests using it instead of blinky if the latter can't be used;
 * I personally believe that synchronization is not an equivalent simple app to blinky due to the aforementioned additional surface area
 * there's precedent for hello_world to simply terminate and not do anything more

So here's an alternative approach I'm happy to run with if that raise no complaints: introduce a new application that is, semantically "blinky but on a serial port", update the documentation to suggest the use of that instead of hello_world where blinky can't be uesd.

Maybe something with a repeating pattern like the old C64 demos? Like "\/" sent very one second? There's a bit more use in slightly longer patterns as they're easier to caught with the debugger though, so maybe a `"Zephyr says 'Hi!'"` line every second?

**Comment by nashif** [comment]
> So here's an alternative approach I'm happy to run with if that raise no complaints: introduce a new application that is, semantically "blinky but on a serial port", update the documentation to suggest the use of that instead of hello_world where blinky can't be uesd.

that works, instead of changing the existing hello_world, you can add a new one implementing similar functionality to blinky for development/debugging needs.

**Comment by gmarull** [comment]
so what about blinky printing every second as well?

**Comment by Flameeyes** [comment]
> so what about blinky printing every second as well?

blinky has a minimum hardware requirement for the led0 alias, or it'll fail to build — which is where the recommendation for hello_world comes from. 

I mean, we could still have it print as well to combine that option, but complicating the sample to make led0 optional doesn't seem like a good idea.

**Comment by Flameeyes** [comment]
Not entirely sure on what's the best practice between re-using the same pull request or separating it, but I think keeping the discussion in one place is worthwhile — happy to submit as a separate pull request if you think it's better.

**Review by gmarull — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Comment by Flameeyes** [comment]
More to the need of not relying on philosophers: as of today philosophers causes a hard fault on the EV11L78A because _something_ is broken in the threading code.

@nashif what's the blocker for having this new sample app?

**Comment by carlescufi** [comment]
@kartben can you take another look?

**Comment by fabiobaltieri** [comment]
@kartben ping

File: samples/basic/greetings/README.rst [hunk 3]
```diff
@@ -0,0 +1,39 @@
+.. _greetings:
+
+Greetings
+#########
```
Comment (kartben) [inline samples/basic/greetings/README.rst:4]:
```suggestion
.. zephyr:code-sample:: greetings
   :name: Greetings
   :relevant-api: thread_apis

   Print greetings to the console once per second.
```

File: samples/basic/greetings/README.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+Overview
+********
+
+A simple sample that can be used with any :ref:`supported board <boards>` and
+prints "Zephyr says 'Hi!'" to the console, once per second.
+
+It's an alternative to :ref:`blinky <blinky-sample>` for boards with no
```
Comment (kartben) [inline samples/basic/greetings/README.rst:12]:
```suggestion
It's an alternative to :zephyr:code-sample:`blinky` for boards with no
```

File: samples/basic/greetings/README.rst [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+controllable LEDs.
+
+Building and Running
+********************
+
+This application can be built and executed on QEMU as follows:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/greetings
```
Comment (kartben) [inline samples/basic/greetings/README.rst:21]:
```suggestion
   :zephyr-app: samples/basic/greetings
```

File: doc/develop/getting_started/index.rst [hunk 6]
```diff
@@ -652,7 +652,7 @@ Build the Blinky Sample
 
    Blinky is compatible with most, but not all, :ref:`boards`. If your board
    does not meet Blinky's :ref:`blinky-sample-requirements`, then
-   :ref:`hello_world` is a good alternative.
+   :ref:`greetings` is a good alternative.
```
Comment (kartben) [inline doc/develop/getting_started/index.rst:655]:
```suggestion
   :zephyr:code-sample:`greetings` is a good alternative.
```

**Review by kartben — CHANGES_REQUESTED** [review]
an actual mistake re: sample path + some requested changes to leverage the recently introduced Zephyr sample Sphinx extension. Thanks @Flameeyes!

**Comment by kartben** [comment]
@Flameeyes will you have cycles to address the opened comments? Thanks!

**Comment by Flameeyes** [comment]
Probably next week, let me see!

**Comment by kartben** [comment]
@Flameeyes thanks! FWIW blinky is now *also* printing to the console, so this might actually not be all that necessary to add this code sample? (PR #63056)

**Comment by Flameeyes** [comment]
Given this started while trying to get blinky to print, if we settled on that I don't need that.

**Pull request closed without merging**